### PR TITLE
[MAP] display location of products on map using pins

### PR DESCRIPTION
**Pull request by cwang12197**

### CODE REVIEW

**Comment by cwang12197** [comment]
1. Created "MarketDetail" component displaying product information and location (in progress) on click from home page 
2. Added markers of product locations onto maps with a skeleton popup modal to display product information 
3. Fixed bug with user error when setting state by defaulting to null and checking for state before loading page

File: food-waste-ui/src/components/App/App.jsx [hunk 1]
```diff
@@ -119,34 +119,43 @@ export function MainApp({
           setCurrentUser(curUser);
 
           let allProducts = response.data.products;
-          //TODO: filter products for search
-          setProducts(allProducts);
+          let newProducts = allProducts.filter((item) =>
+            item.name.toLowerCase().includes(searchTerm.toLowerCase())
```
Comment (jennyhan1999) [inline food-waste-ui/src/components/App/App.jsx:123]: Something to think about is user errors whenever theres a form/input. What happens if there is extra trailing/leading whitespace or if the user spells the searchTerm incorrectly? A stretch could be to add suggested search results based on the user input with errors.

File: food-waste-ui/src/components/App/App.jsx [hunk 2]
```diff
@@ -119,34 +119,43 @@ export function MainApp({
...
+          setProducts(newProducts);
           const userProducts = allProducts.filter(
             (item) => item.user.objectId == curUser.objectId
           ); //the thing with setState it is async, if setting state and doing based off state after might not update
           setMyProducts(userProducts);
-          setIsLoading(false);
         });
       })
       .catch((err) => {
         console.log(err);
```
Comment (jennyhan1999) [inline food-waste-ui/src/components/App/App.jsx:133]: When you have not implemented something yet you can add a comment that looks like
// TODO @cwang12197 : Add error logging and error screen

File: food-waste-ui/src/components/App/App.jsx [hunk 3]
```diff
@@ -119,34 +119,43 @@ export function MainApp({
...
 
   const handleOnToggle = () => {
     setIsOpen(!isOpen);
   };
 
+  const handleSearchChange = (event) => {
+    event.preventDefault();
+    setSearchTerm(event.target.value);
+    console.log("search", searchTerm);
+    console.log("products", products);
```
Comment (jennyhan1999) [inline food-waste-ui/src/components/App/App.jsx:145]: nit: clean up console.log

**Review by jennyhan1999 — COMMENTED** [review]
Overall looks good! left some comments

File: food-waste-express-api/routes/login.js [hunk 4]
```diff
@@ -2,16 +2,37 @@ const express = require("express")
...
 
-
+router.put('/', async (req, res) => {
+  const userId = req.body.objectId;
+  const User = new Parse.User()
+  const query = new Parse.Query(User)
+  try {
+    let user = await query.get(userId)
+    console.log("USER FJWE", user)
+    console.log("locationijrf", req.body.location)
```
Comment (jennyhan1999) [inline food-waste-express-api/routes/login.js:23]: Clean up debugging code before submitting

File: food-waste-express-api/routes/login.js [hunk 5]
```diff
@@ -2,16 +2,37 @@ const express = require("express")
...
+    await user.save()
+    try {
+      // Saves the user with the updated data
+      let response = await user.save();
+      console.log('Updated user', response);
+    } catch (error) {
+      console.error('Error while updating user', error);
+    }
+  } catch (error) {
+    console.error('Error while retrieving user', error);
```
Comment (jennyhan1999) [inline food-waste-express-api/routes/login.js:34]: Add a TODO, see below for template:

// TODO @owner: Short description of task

File: food-waste-ui/src/components/Home/Home.jsx [hunk 6]
```diff
@@ -4,13 +4,25 @@ import MarketGrid from "../MarketGrid/MarketGrid";
 // import SearchBar from "../SearchBar/SearchBar";
```
Comment (jennyhan1999) [inline food-waste-ui/src/components/Home/Home.jsx:4]: Nit: remove

File: food-waste-ui/src/components/Map/Map.jsx [hunk 7]
```diff
@@ -2,51 +2,155 @@ import * as React from "react";
...
 
-  return !isLoaded ? (
-    <Skeleton />
-  ) : (
+  const calculateRoute = async () => {
+    if (origin == "" || !destination) {
+      return;
+    }
+    const directionService = new google.maps.DirectionsService();
+    console.log("destination in service", destination);
```
Comment (jennyhan1999) [inline food-waste-ui/src/components/Map/Map.jsx:76]: Nit: clean up

File: food-waste-ui/src/components/MarketDetail/MarketDetail.jsx [hunk 8]
```diff
@@ -1,3 +1,46 @@
 import * as React from "react";
+import axios from "axios";
```
Comment (jennyhan1999) [inline food-waste-ui/src/components/MarketDetail/MarketDetail.jsx:2]: You could split this part into 2 diffs, 1 for using the google map api and one for this market detail

**Comment by cwang12197** [comment]
**[DESCRIPTION]**

- fixed marker error 
- successfully displays direction and distance/duration
- successfully updates center of map based on user location when logging in 
- clears route using "x" button 
- when searching for items, map also updates so only the search result products are shown :)

[WORKING ON]

- reverse geocoding addresses/consider changing the input values to something else instead of "My Location" and "Product Location" 
- fixing the UI and adding pop-ups to the markers when clicked on 
Screen Shot 2022-07-21 at 12 24 25 AM
Screen Shot 2022-07-21 at 12 25 15 AM

**Review by jennyhan1999 — APPROVED** [review]
LGTM!

**Comment by cwang12197** [comment]
**[DESCRIPTION]**

- fixed a warning when loading maps that states center as "NaN"
- display maps and product card when clicking into the product 
- reverse geocoded the locations so they show up as proper addresses instead of coordinates (as close as possible given walking)

**[WORKING ON]** 

- layout for product detail, popup, map view 
- setting up directions to automatically load when in product detail

**Comment by cwang12197** [comment]
**[DESCRIPTION]**
- automatically display user location on maps (to be customized later) and product location when clicked on in product detail
This is a screenshot of the map page, with only the user location displayed:
Screen Shot 2022-07-22 at 3 30 24 PM

This is a screenshot of the product detail page (of rice), with both the user and product location displayed so the route can be calculated after clicking on the button: 
Screen Shot 2022-07-22 at 3 31 09 PM

**Review by jennyhan1999 — APPROVED** [review]
looks good!

**Pull request merged**